File: _98_ValidateBinarySearchTree.py
from typing import List
import functools
from pytest import mark

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        nodeList = []
        
        def inorder(node):
            if not node:
                return
            inorder(node.left)
            nodeList.append(node.val)
            inorder(node.right)
            
        inorder(root)
        for i in range(len(nodeList) - 1):
            if nodeList[i] >= nodeList[i+1]:
                return False
        return True

File: _98_ValidateBinarySearchTree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isValidBST(TreeNode root) {
        if (root == null) return true;
        if (!compareNode(root.val, true, root.left)
            || !compareNode(root.val, false, root.right)) {
                return false;
        }
        return isValidBST(root.left) 
                && isValidBST(root.right);
    }

    public boolean compareNode(int val, boolean left, TreeNode node) {
        if (node == null) return true;
        if ((left && node.val >= val)
            || (!left && node.val <= val)) {
            return false;
        } else {
            return compareNode(val, left, node.left)
                    && compareNode(val, left, node.right);
        }
    }
}